File: demo_core_av_2/app/src/main/java/com/bandyer/demo_core_av_2/design/bottom_sheet/picker/BottomListPickerItem.java
/*
 * Copyright (C) 2018 Bandyer S.r.l. All Rights Reserved.
 * See LICENSE.txt for licensing information
 */

package com.bandyer.demo_core_av_2.design.bottom_sheet.picker;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bandyer.demo_core_av_2.R;
import com.mikepenz.fastadapter.FastAdapter;
import com.mikepenz.fastadapter.items.AbstractItem;
import com.mikepenz.fastadapter.listeners.ClickEventHook;

import net.igenius.customcheckbox.CustomCheckBox;

import java.util.List;
import java.util.Set;

import butterknife.BindView;
import butterknife.ButterKnife;


/**
 * @author kristiyan
 **/

public class BottomListPickerItem<T> extends AbstractItem<BottomListPickerItem, BottomListPickerItem.Holder<T>> {

    private T item;

    @NonNull
    @Override
    public Holder<T> getViewHolder(@NonNull View v) {
        return new BottomListPickerItem.Holder<>(v);
    }

    @Override
    public int getType() {
        return R.id.picker_item;
    }

    @Override
    public int getLayoutRes() {
        return R.layout.bottom_list_picker_item;
    }

    public interface Delegate<T> {
        void onClicked(int adapterPosition, T item);
    }

    public BottomListPickerItem<T> setItem(T item) {
        this.item = item;
        return this;
    }

    public T getItem() {
        return item;
    }

    @Override
    public void bindView(@NonNull Holder<T> holder, @NonNull List<Object> payloads) {
        super.bindView(holder, payloads);
        holder.checked.setChecked(isSelected());
    }

    public static class Holder<T> extends FastAdapter.ViewHolder<BottomListPickerItem<T>> {

        @BindView(R.id.picker_item)
        LinearLayout picker_item;

        @BindView(R.id.item)
        TextView item;

        @BindView(R.id.checked)
        CustomCheckBox checked;

        Holder(View view) {
            super(view);
            ButterKnife.bind(this, view);
        }

        @Override
        public void bindView(@NonNull BottomListPickerItem<T> listItem, @NonNull List<Object> payloads) {
            item.setText(listItem.item.toString());
        }

        @Override
        public void unbindView(@NonNull BottomListPickerItem<T> item) {

        }

    }

    public static class PickerItemClick<T> extends ClickEventHook<BottomListPickerItem<T>> {

        BottomListPickerItem.Delegate<T> delegate;

        PickerItemClick(BottomListPickerItem.Delegate<T> delegate) {
            this.delegate = delegate;
        }

        @Override
        public View onBind(@NonNull RecyclerView.ViewHolder viewHolder) {
            if (viewHolder instanceof BottomListPickerItem.Holder) {
                return ((BottomListPickerItem.Holder) viewHolder).picker_item;
            }
            return null;
        }

        @Override
        @SuppressWarnings("deprecation")
        public void onClick(@NonNull View v, int position, @NonNull FastAdapter<BottomListPickerItem<T>> fastAdapter, @NonNull BottomListPickerItem<T> item) {
            if (item.isSelected())
                return;

            Set<Integer> selections = fastAdapter.getSelections();

            if (selections.isEmpty()) {
                fastAdapter.select(position);
                delegate.onClicked(position, item.getItem());
                return;
            }

            int selectedPosition = selections.iterator().next();
            fastAdapter.deselect(selectedPosition);
            fastAdapter.notifyItemChanged(selectedPosition);

            fastAdapter.select(position);
            delegate.onClicked(position, item.getItem());
        }
    }

    public static class GroupPickerItemClick<T> extends ClickEventHook<BottomListPickerItem<T>> {

        @Override
        public View onBind(@NonNull RecyclerView.ViewHolder viewHolder) {
            if (viewHolder instanceof BottomListPickerItem.Holder) {
                return ((BottomListPickerItem.Holder) viewHolder).picker_item;
            }
            return null;
        }

        @Override
        @SuppressWarnings("deprecation")
        public void onClick(@NonNull View v, int position, @NonNull FastAdapter<BottomListPickerItem<T>> fastAdapter, @NonNull BottomListPickerItem<T> item) {
            if (item.isSelected())
                return;

            Set<Integer> selections = fastAdapter.getSelections();

            if (selections.isEmpty()) {
                fastAdapter.select(position);
                return;
            }

            for (Integer selectedPosition : selections) {
                if (item.getTag().equals(fastAdapter.getItem(selectedPosition).getTag())) {
                    fastAdapter.deselect(selectedPosition);
                    fastAdapter.notifyItemChanged(selectedPosition);
                }
            }

            fastAdapter.select(position);
        }
    }
}

File: demo_core_av/app/src/main/java/com/bandyer/demo_core_av/design/bottom_sheet/picker/BottomListPickerItem.kt
/*
 * Copyright (C) 2018 Bandyer S.r.l. All Rights Reserved.
 * See LICENSE.txt for licensing information
 */
package com.bandyer.demo_core_av.design.bottom_sheet.picker

import android.view.View
import androidx.recyclerview.widget.RecyclerView
import com.bandyer.demo_core_av.R
import com.mikepenz.fastadapter.FastAdapter
import com.mikepenz.fastadapter.IItem
import com.mikepenz.fastadapter.items.AbstractItem
import com.mikepenz.fastadapter.listeners.ClickEventHook
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.bottom_list_picker_item.view.*

/**
 * @author kristiyan
 */
class BottomListPickerItem<T> : AbstractItem<BottomListPickerItem<*>?, BottomListPickerItem.Holder<T>>() {

    var item: T? = null
        private set

    override fun getViewHolder(v: View): Holder<T> = Holder(v)

    override fun getType(): Int = R.id.picker_item

    override fun getLayoutRes(): Int = R.layout.bottom_list_picker_item

    interface Delegate<T> {
        fun onClicked(adapterPosition: Int, item: T?)
    }

    fun setItem(item: T): BottomListPickerItem<T> {
        this.item = item
        return this
    }

    override fun bindView(holder: Holder<T>, payloads: List<Any>) {
        super.bindView(holder, payloads)
        holder.containerView.checked!!.isChecked = isSelected
    }

    class Holder<T> internal constructor(override val containerView: View) : FastAdapter.ViewHolder<BottomListPickerItem<T>>(containerView), LayoutContainer {

        override fun bindView(listItem: BottomListPickerItem<T>, payloads: List<Any>) {
            containerView.item!!.text = listItem.item.toString()
        }

        override fun unbindView(item: BottomListPickerItem<T>) {}
    }

    class PickerItemClick<T> internal constructor(val delegate: Delegate<T>) : ClickEventHook<BottomListPickerItem<T>>() {

        override fun onBind(viewHolder: RecyclerView.ViewHolder): View? {
            return if (viewHolder is Holder<*>) viewHolder.containerView.findViewById(R.id.picker_item) else null
        }

        override fun onClick(v: View, position: Int, fastAdapter: FastAdapter<BottomListPickerItem<T>>, item: BottomListPickerItem<T>) {
            if (item.isSelected) return
            val selections = fastAdapter.selections
            if (selections.isEmpty()) {
                fastAdapter.select(position)
                delegate.onClicked(position, item.item)
                return
            }
            val selectedPosition = selections.iterator().next()
            fastAdapter.deselect(selectedPosition)
            fastAdapter.notifyItemChanged(selectedPosition)
            fastAdapter.select(position)
            delegate.onClicked(position, item.item)
        }

    }


    class GroupPickerItemClick : ClickEventHook<IItem<*, *>?>() {
        override fun onBind(viewHolder: RecyclerView.ViewHolder): View? {
            return if (viewHolder is Holder<*>) viewHolder.containerView.findViewById(R.id.picker_item) else null
        }

        override fun onClick(v: View, position: Int, fastAdapter: FastAdapter<IItem<*,*>?>, item: IItem<*,*>?) {
            if (item?.isSelected != false) return
            val selections = fastAdapter.selections
            if (selections.isEmpty()) {
                fastAdapter.select(position)
                return
            }
            selections.forEach { selectedPosition ->
                if (item.tag == fastAdapter.getItem(selectedPosition!!)?.tag) {
                    fastAdapter.deselect(selectedPosition)
                    fastAdapter.notifyItemChanged(selectedPosition)
                }
            }
            fastAdapter.select(position)
        }
    }
}